text/0062-controlling-pii-and-credentials-in-sd-ks.md

```diff
@@ -0,0 +1,207 @@
+                l[i] = self.scrub_dict(v)
+            elif isinstance(v, list):
+                l[i] = self.scrub_list(v)
+          return l
+
+    def scrub_dict(self, d):
+        for k, v in d.items():
+            if k in self.denylist:
+                d[k] = "[Filtered]"  # Use `AnnotatedValue` here for _meta
+            elif isinstance(v, dict):
+                d[k] = self.scrub_dict(v)
+            elif isinstance(v, list):
+                d[k] = self.scrub_list(v)
+        return d
+
+    def scrub_event(self, event):
+        return self.scrub_dict(event)
+```
+
+Here we would walk the whole event dict and whenever we encounter a key from the denylist, we would replace the value with `[Filtered]`.
+We can optionally annotate the values so that the UI is transparent about the filtering.  
+This allows a centralized place to scrub stuff regardless of where in the SDK it originates from.
+
+This basically mirrors what relay does but with comparatively less complexity in how we scrub.  
+Relay does a bunch of regex matching and such which would be a lot of added complexity to add to each SDK.
+
+Note that the actual implementation of this recursion might have differences based on platform and the typing of `Event`. The above is just pseudo-code.  
+
+## B: Scrub pre-defined set of interfaces
+
+Instead of recursing through the whole payload, we can also manually keep track of interfaces we expect to see PII in and just scrub those.  
+The advantage here is that it also acts as a form of self-documentation for sources of leakage.  
+The implementation in this case would look something like the following.
+
+```python
+class EventScrubber:
+    def scrub_dict(self, d):
+        for k, v in d.items():
+            if k in self.denylist:
+                d[k] = "[Filtered]"
+
+    def scrub_event(self, event):
+        if "request" in event:
+            if "headers" in event["request"]:
+                self.scrub_dict(event["request"]["headers"])  # or event.request.headers if you have interfaces in your language
+
+        if "spans" in event:
+            for span in event["spans"]:
+                if "data" in span:
+                    self.scrub_dict(span["data"])
+
+        # ... all other such fields exhaustively
+        return event
+```
+
+For completeness, here's a list of interfaces from `sentry-python` currently potentially leaking PII.
+
+* `event.request.headers`
+* `event.request.cookies`
+* `event.request.data`
+* `event.logentry.params`
+* `event.user`
+* `event.breadcrumbs.values -> value.data`
+* `event.exception.values -> value.stacktrace.frames -> frame.vars`
+* `event.spans -> span.data`
+
+## (Optional) C: Add separate security denylist
+
+A related proposal is to split the `denylist` into `pii_denylist` and `security_denylist` where stuff in `security_denylist` is **always** scrubbed
+irrespective of `send_default_pii` and stuff in `pii_denylist` is only scrubbed if `send_default_pii` is `False`.
```

Comment:
I am for option A) (because it is secure by default and will not rot as fast)

But with two deny lists for PII and security data. 

To make it impossible to send session data by setting `sendDefaultPII` to true. It could be though that some people are depending on having session data for debugging, but if a lot of people complain we can introduce `sendDefaultSecurityData`

Reply:
+1

A separation between the two would be great here. It would significantly reduce the risks one has to consider when toggling the `sendDefaultPII` option.

Some may gravitate to using a sensitive value like a session cookie to track through their apps, but there are more secure ways of doing this with an identifier not inherently tied to auth.

Reply:
:O thanks for stopping by on my RFC @mdtro! <3